File: src/main/java/parser/ast/expressions/BinaryExpression.java
package parser.ast.expressions;

import stdlib.ArrayValue;
import stdlib.NumberValue;
import stdlib.StringValue;
import stdlib.Value;

public final class BinaryExpression implements Expression {

    private Expression expr1, expr2;
    private char operation;
    private int _opPosStr;
    private int _opPosLine;


    public BinaryExpression(char operation, Expression expr1, Expression expr2,  int opPosStr, int opPosLine) {
        this.expr1 = expr1;
        this.expr2 = expr2;
        this.operation = operation;
        _opPosStr = opPosStr;
        _opPosLine = opPosLine;
    }

    @Override
    public Value eval() {
        final var value1 = expr1.eval();
        final var value2 = expr2.eval();
        if ((value1 instanceof StringValue) || (value1 instanceof ArrayValue)){
            final var string1 = value1.asString();
            switch (operation){
                case '*' : {
                    final var iterations = (int) value2.asNumber();
                    final var buffer = new StringBuilder();
                    for (var i = 0; i < iterations; i++){
                        buffer.append(string1);
                    }
                    return new StringValue(buffer.toString());
                }
                case '+' :
                default:
                    return new StringValue(string1 + value2.asString());
            }
        }

        final var number1 = value1.asNumber();
        final var number2 = value2.asNumber();
        switch (operation){
            case '-' : return new NumberValue(number1 - number2);
            case '*' : return new NumberValue(number1 * number2);
            case '/' : return new NumberValue(number1 / number2);
            case '+' :
            default: return new NumberValue(number1 + number2);
        }
    }

    @Override
    public String toString() {
        return String.format("<BinaryExpression>%s<OperationExpression operation=\"%c\" str=\"%d\" line=\"%d\"/>%s</BinaryExpression>", expr1, operation, _opPosStr, _opPosLine, expr2);

    }
}

File: src/main/java/parser/ast/expressions/BinaryExpression.kt
package parser.ast.expressions

import stdlib.ArrayValue
import stdlib.NumberValue
import stdlib.StringValue
import stdlib.Value

class BinaryExpression(private val operation: Char, private val expr1: Expression, private val expr2: Expression, private val _opPosStr: Int, private val _opPosLine: Int) : Expression {

    override fun eval(): Value {
        val value1 = expr1.eval()
        val value2 = expr2.eval()
        if (value1 !is StringValue && value1 !is ArrayValue) {
            val number1 = value1.asNumber()
            val number2 = value2.asNumber()
            return when (operation) {
                '-' -> NumberValue(number1 - number2)
                '*' -> NumberValue(number1 * number2)
                '/' -> NumberValue(number1 / number2)
                '+' -> NumberValue(number1 + number2)
                else -> NumberValue(number1 + number2)
            }
        }
        val string1 = value1.asString()
        return when (operation) {
                '*' -> {
                    val iterations = value2.asNumber().toInt()
                    val buffer = StringBuilder()
                    for (i in 0 until iterations) {
                        buffer.append(string1)
                    }
                    StringValue(buffer.toString())
                }
                '+' -> StringValue(string1 + value2.asString())
                else -> StringValue(string1 + value2.asString())
            }
    }

    override fun toString(): String {
        return String.format("<BinaryExpression>%s<OperationExpression operation=\"%c\" str=\"%d\" line=\"%d\"/>%s</BinaryExpression>", expr1, operation, _opPosStr, _opPosLine, expr2)

    }
}
